ddl/ddl_api.go

```diff
@@ -221,18 +262,69 @@ func (d *ddl) ModifySchemaSetTiFlashReplica(ctx sessionctx.Context, stmt *ast.Al
 	if total == 0 {
 		return infoschema.ErrEmptyDatabase.GenWithStack("Empty database '%v'", dbName.O)
 	}
-	err := checkTiFlashReplicaCount(ctx, tiflashReplica.Count)
+	err := checkTiFlashReplicaCount(sctx, tiflashReplica.Count)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
+	var originVersion int64 = 0
+	var pendingCount uint32 = 0
+	forceCheck := false
+
+	logutil.BgLogger().Info("start batch add TiFlash replicas", zap.Int("total", total), zap.Int64("schemaID", dbInfo.ID))
 	for _, tbl := range dbInfo.Tables {
+		select {
+		case <-ctx.Done():
+			logutil.BgLogger().Info("abort batch add TiFlash replica", zap.Int64("schemaID", dbInfo.ID))
+			return nil
+		default:
+		}
+
 		tbReplicaInfo := tbl.TiFlashReplica
 		if !shouldModifyTiFlashReplica(tbReplicaInfo, tiflashReplica) {
 			logutil.BgLogger().Info("skip processing schema table", zap.Int64("tableID", tbl.ID), zap.Int64("schemaID", dbInfo.ID), zap.String("tableName", tbl.Name.String()), zap.String("schemaName", dbInfo.Name.String()))
 			skip += 1
 			continue
 		}
+
+		configRetry := tiflashCheckPendingTablesRetry
+		configWaitTime := tiflashCheckPendingTablesWaitTime
+		failpoint.Inject("FastFailCheckTiFlashPendingTables", func(value failpoint.Value) {
+			configRetry = value.(int)
+			configWaitTime = time.Second
+		})
+		// When handled some tables, we need to wait some tables become available.
+		if (succ+fail)%tiflashCheckPendingTablesTick == 0 || forceCheck {
```

Comment:
Maybe these codes should be moved before `shouldModifyTiFlashReplica`. Also, it gets complicated, `tiflashCheckPendingTablesTick` and `forceCheck` is getting this loop more and more complex.

I would suggest `succ+fail > threshold`, in which `threshold` is got before the whole loop.

Reply:
Modified according to suggestions, except:
1. Move these code outside the loop instead of just before `shouldModifyTiFlashReplica`, in order to make the loop smaller.
2. I find it hard to use `succ+fail > threshold`, because `succ+fail` is global, and is expected to be larger than threshold.
3. `forceCheck` is used to enforce check after execution of a probing DDL.